File: ios_app/iphone/ReadmeSheet.swift
import SwiftUI
import shared

private let pTextLineHeight = 3.2

extension Fs {

    func ReadmeSheet__open(
        defaultItem: ReadmeSheetVm.DefaultItem = .basics
    ) {
        self.show { isPresented in
            ReadmeSheet(
                isPresented: isPresented,
                defaultItem: defaultItem
            )
        }
    }
}

private struct ReadmeSheet: View {
    
    @Binding private var isPresented: Bool
    
    @State private var vm: ReadmeSheetVm
    
    init(
        isPresented: Binding<Bool>,
        defaultItem: ReadmeSheetVm.DefaultItem
    ) {
        _isPresented = isPresented
        vm = ReadmeSheetVm(defaultItem: defaultItem)
    }
    
    var body: some View {
        
        VMView(vm: vm, stack: .VStack()) { state in
            
            Fs__HeaderTitle(
                title: state.title,
                scrollToHeader: 0,
                onClose: {
                    isPresented = false
                }
            )
            
            TabView(tabUi: state.tabUi)
                .id("tab_\(state.tabUi.id)")
            
            DividerBg()
                .padding(.horizontal, H_PADDING)
            
            HStack {
                
                ForEachIndexed(state.tabsUi) { idx, tabUi in
                    TabBarItemView(
                        title: tabUi.title,
                        isActive: state.tabUi.id == tabUi.id,
                        onClick: {
                            vm.setTabUi(tabUi: tabUi)
                        }
                    )
                    Padding(horizontal: 6)
                }
                
                Spacer()
            }
            .padding(.top, 9)
            .padding(.bottom, 4)
            .padding(.horizontal, H_PADDING)
            .safeAreaPadding(.bottom)
        }
    }
}

///

private let tabBarItemViewShape = RoundedRectangle(cornerRadius: 10, style: .continuous)

private struct TabBarItemView: View {
    
    let title: String
    let isActive: Bool
    let onClick: () -> Void
    
    var body: some View {
        
        Button(
            action: {
                onClick()
            },
            label: {
                Text(title)
                    .padding(.horizontal, 9)
                    .padding(.top, 3)
                    .padding(.bottom, 3 + halfDpFloor)
                    .foregroundColor(isActive ? c.white : c.text)
                    .background(tabBarItemViewShape.fill(isActive ? c.blue : c.transparent))
                    .font(.system(size: 15, weight: .medium))
            }
        )
    }
}

private struct ImagePreviewsView: View {
    
    let images: [String]
    
    @EnvironmentObject private var fs: Fs
    
    var body: some View {
        
        ScrollView(.horizontal, showsIndicators: false) {
            
            HStack {
                
                Padding(horizontal: 10)
                
                ForEachIndexed(images) { _, item in
                    
                    Button(
                        action: {
                            fs.show { isSliderPresented in
                                ImagesSlider(
                                    isPresented: isSliderPresented,
                                    images: images
                                )
                            }
                        },
                        label: {
                            Image(item)
                                .resizable()
                                .aspectRatio(contentMode: .fit)
                                .cornerRadius(16)
                                .shadow(color: c.text, radius: onePx)
                                .frame(height: 350)
                                .padding(.horizontal, 6)
                                .padding(.vertical, 4) // Paddings for shadow radius
                        }
                    )
                }
                
                Padding(horizontal: 10)
            }
        }
        .padding(.top, 20)
    }
}

private struct ImagesSlider: View {
    
    @Binding var isPresented: Bool
    let images: [String]
    
    var body: some View {
        
        VStack {
            
            ScrollView(.horizontal, showsIndicators: false) {
                
                HStack {
                    
                    Padding(horizontal: 10)
                    
                    ForEachIndexed(images) { _, item in
                        
                        Image(item)
                            .resizable()
                            .aspectRatio(contentMode: .fit)
                            .cornerRadius(16)
                            .shadow(color: c.text, radius: onePx)
                            .frame(height: .infinity)
                            .padding(.horizontal, 12)
                        // Paddings for shadow radius
                            .padding(.vertical, 8)
                    }
                    
                    Padding(horizontal: 10)
                }
            }
            
            HStack {
                Spacer()
                Button(
                    action: {
                        isPresented = false
                    },
                    label: {
                        Text("close")
                            .foregroundColor(c.text)
                    }
                )
                .padding(.top, 12)
                .padding(.bottom, 12)
                .padding(.trailing, 36)
            }
        }
        .safeAreaPadding(.vertical)
    }
}

private struct TabView: View {
    
    let tabUi: ReadmeSheetVm.TabUi
    @State private var scroll = 0
    
    var body: some View {
        
        VStack {
            
            DividerBgScroll(scrollToHeader: scroll)
            
            ScrollViewWithVListener(showsIndicators: true, vScroll: $scroll) {
                
                VStack {
                    
                    let paragraphs = tabUi.paragraphs
                    
                    ForEachIndexed(paragraphs) { idx, paragraph in
                        
                        let prevP: ReadmeSheetVm.Paragraph? = (idx == 0) ? nil : paragraphs[idx - 1]
                        
                        if let paragraph = paragraph as? ReadmeSheetVm.ParagraphTitle {
                            let paddingTop: CGFloat = {
                                guard let prevP = prevP else {
                                    return 12
                                }
                                if prevP.isSlider {
                                    return 40
                                }
                                if prevP is ReadmeSheetVm.ParagraphText {
                                    return 39
                                }
                                fatalError()
                            }()
                            HStack {
                                Text(paragraph.text)
                                    .foregroundColor(c.text)
                                    .font(.system(size: Fs__TITLE_FONT_SIZE, weight: Fs__TITLE_FONT_WEIGHT))
                                    .padding(.top, paddingTop)
                                    .padding(.horizontal, H_PADDING)
                                Spacer()
                            }
                        } else if let paragraph = paragraph as? ReadmeSheetVm.ParagraphText {
                            let paddingTop: CGFloat = {
                                guard let prevP = prevP else {
                                    return 14
                                }
                                if prevP.isSlider {
                                    return 16
                                }
                                if prevP is ReadmeSheetVm.ParagraphTitle {
                                    return 17
                                }
                                if prevP is ReadmeSheetVm.ParagraphText {
                                    return 16
                                }
                                if prevP is ReadmeSheetVm.ParagraphTextHighlight {
                                    return 20
                                }
                                fatalError()
                            }()
                            HStack {
                                Text(paragraph.text)
                                    .foregroundColor(c.text)
                                    .padding(.top, paddingTop)
                                    .padding(.horizontal, H_PADDING)
                                    .lineSpacing(pTextLineHeight)
                                Spacer()
                            }
                        } else if let paragraph = paragraph as? ReadmeSheetVm.ParagraphTextHighlight {
                            let paddingTop: CGFloat = {
                                guard let prevP = prevP else {
                                    fatalError()
                                }
                                if prevP.isSlider {
                                    return 24
                                }
                                if prevP is ReadmeSheetVm.ParagraphText {
                                    return 23
                                }
                                fatalError()
                            }()
                            HStack {
                                Text(paragraph.text)
                                    .foregroundColor(c.text)
                                    .padding(.vertical, 13)
                                    .padding(.horizontal, 14)
                                    .lineSpacing(pTextLineHeight)
                                Spacer()
                            }
                            .background(squircleShape.fill(c.blue))
                            .padding(.top, paddingTop)
                            .padding(.horizontal, H_PADDING - 2)
                        } else if let paragraph = paragraph as? ReadmeSheetVm.ParagraphAskAQuestion {
                            AskAQuestionButtonView(
                                subject: paragraph.subject,
                                isFirst: true,
                                isLast: true,
                                bgColor: c.fg,
                                withTopDivider: false
                            )
                            .padding(.top, 24)
                        } else if paragraph is ReadmeSheetVm.ParagraphTimerTypical {
                            ImagePreviewsView(
                                images: [
                                    "readme_timer_1"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphTimerMyActivities {
                            ImagePreviewsView(
                                images: [
                                    "readme_activities_1"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphTimerCharts {
                            ImagePreviewsView(
                                images: [
                                    "readme_chart_1",
                                    "readme_chart_2",
                                    "readme_chart_3"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphTimerPractice1 {
                            ImagePreviewsView(
                                images: [
                                    "readme_timer_practice_1",
                                    "readme_timer_practice_2",
                                    "readme_timer_practice_3",
                                    "readme_timer_practice_4"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphTimerPractice2 {
                            ImagePreviewsView(
                                images: [
                                    "readme_timer_practice_5",
                                    "readme_chart_2",
                                    "readme_chart_3"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphRepeatingsMy {
                            ImagePreviewsView(
                                images: [
                                    "readme_repeatings_1"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphRepeatingsToday {
                            ImagePreviewsView(
                                images: [
                                    "readme_repeatings_2"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphRepeatingsPractice1 {
                            ImagePreviewsView(
                                images: [
                                    "readme_repeating_practice_1",
                                    "readme_repeating_practice_2",
                                    "readme_repeating_practice_3"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphRepeatingsPractice2 {
                            ImagePreviewsView(
                                images: [
                                    "readme_repeating_practice_4",
                                    "readme_repeating_practice_5"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphChecklistsExamples {
                            ImagePreviewsView(
                                images: [
                                    "readme_checklists_1",
                                    "readme_checklists_2",
                                    "readme_checklists_3"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphChecklistsPractice1 {
                            ImagePreviewsView(
                                images: [
                                    "readme_checklists_practice_1",
                                    "readme_checklists_practice_2",
                                    "readme_checklists_practice_3",
                                    "readme_checklists_practice_4",
                                    "readme_checklists_practice_5",
                                    "readme_checklists_practice_6",
                                    "readme_checklists_practice_7"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphChecklistsPractice2 {
                            ImagePreviewsView(
                                images: [
                                    "readme_checklists_practice_8",
                                    "readme_checklists_practice_9"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphPomodoroExamples {
                            ImagePreviewsView(
                                images: [
                                    "readme_pomodoro_1",
                                    "readme_pomodoro_2",
                                    "readme_pomodoro_3",
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphGoalsExamples {
                            ImagePreviewsView(
                                images: [
                                    "readme_goals_1"
                                ]
                            )
                        } else if paragraph is ReadmeSheetVm.ParagraphCalendarExamples {
                            ImagePreviewsView(
                                images: [
                                    "readme_calendar_1",
                                    "readme_calendar_2"
                                ]
                            )
                        } else {
                            fatalError()
                        }
                    }
                    
                    ZStack {
                    }
                    .padding(.top, 28)
                }
            }
        }
    }
}

File: android_app/src/main/java/me/timeto/app/ui/ReadmeSheet.kt
package me.timeto.app.ui

import androidx.compose.animation.EnterTransition
import androidx.compose.animation.ExitTransition
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.VisibilityThreshold
import androidx.compose.animation.core.spring
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.*
import me.timeto.app.R
import me.timeto.shared.ColorRgba
import me.timeto.shared.vm.ReadmeSheetVM

private val hPadding = MyListView.PADDING_OUTER_HORIZONTAL

private val imagesHBetween = 4.dp
private val imagesHBlock = 14.dp
private val imagesShape = SquircleShape(len = 50f)

private val pTextLineHeight = 22.sp

@Composable
fun ReadmeSheet(
    layer: WrapperView.Layer,
) {

    val (_, state) = rememberVM { ReadmeSheetVM() }

    VStack(
        modifier = Modifier
            .background(c.sheetBg)
    ) {

        val scrollState = rememberScrollState()

        Sheet__HeaderView(
            title = state.title,
            scrollState = scrollState,
            bgColor = c.sheetBg,
        )

        VStack(
            modifier = Modifier
                .verticalScroll(state = scrollState)
                .padding(bottom = 20.dp)
                .weight(1f),
        ) {

            state.paragraphs.forEach { paragraph ->

                when (paragraph) {

                    is ReadmeSheetVM.Paragraph.Title -> PTitleView(paragraph.text)

                    is ReadmeSheetVM.Paragraph.Subtitle -> {
                        Text(
                            text = paragraph.text,
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(horizontal = hPadding)
                                .padding(top = 36.dp),
                            color = c.white,
                            fontWeight = FontWeight.Bold,
                            fontSize = 20.sp,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.Text -> PTextView(paragraph.text)

                    is ReadmeSheetVM.Paragraph.TextHighLight -> PTextHighLightView(paragraph.text)

                    is ReadmeSheetVM.Paragraph.ListDash -> PListDashedView(paragraph.items)

                    is ReadmeSheetVM.Paragraph.TimerTypical -> {
                        ImagePreviewsView(
                            R.drawable.readme_timer_1,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.TimerMyActivities -> {
                        ImagePreviewsView(
                            R.drawable.readme_activities_1,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.TimerCharts -> {
                        ImagePreviewsView(
                            R.drawable.readme_chart_1,
                            R.drawable.readme_chart_2,
                            R.drawable.readme_chart_3,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.TimerPractice1 -> {
                        ImagePreviewsView(
                            R.drawable.readme_timer_practice_1,
                            R.drawable.readme_timer_practice_2,
                            R.drawable.readme_timer_practice_3,
                            R.drawable.readme_timer_practice_4,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.TimerPractice2 -> {
                        ImagePreviewsView(
                            R.drawable.readme_timer_practice_5,
                            R.drawable.readme_chart_2,
                            R.drawable.readme_chart_3,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.RepeatingsMy -> {
                        ImagePreviewsView(
                            R.drawable.readme_repeatings_1,
                        )
                    }


                    is ReadmeSheetVM.Paragraph.RepeatingsToday -> {
                        ImagePreviewsView(
                            R.drawable.readme_repeatings_2,
                        )
                    }


                    is ReadmeSheetVM.Paragraph.RepeatingsPractice1 -> {
                        ImagePreviewsView(
                            R.drawable.readme_repeating_practice_1,
                            R.drawable.readme_repeating_practice_2,
                            R.drawable.readme_repeating_practice_3,
                        )
                    }


                    is ReadmeSheetVM.Paragraph.RepeatingsPractice2 -> {
                        ImagePreviewsView(
                            R.drawable.readme_repeating_practice_4,
                            R.drawable.readme_repeating_practice_5,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.ChecklistsExamples -> {
                        ImagePreviewsView(
                            R.drawable.readme_checklists_1,
                            R.drawable.readme_checklists_2,
                            R.drawable.readme_checklists_3,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.ChecklistsPractice1 -> {
                        ImagePreviewsView(
                            R.drawable.readme_checklists_practice_1,
                            R.drawable.readme_checklists_practice_2,
                            R.drawable.readme_checklists_practice_3,
                            R.drawable.readme_checklists_practice_4,
                            R.drawable.readme_checklists_practice_5,
                            R.drawable.readme_checklists_practice_6,
                            R.drawable.readme_checklists_practice_7,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.ChecklistsPractice2 -> {
                        ImagePreviewsView(
                            R.drawable.readme_checklists_practice_8,
                            R.drawable.readme_checklists_practice_9,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.GoalsExamples -> {
                        ImagePreviewsView(
                            R.drawable.readme_goals_1,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.CalendarExamples -> {
                        ImagePreviewsView(
                            R.drawable.readme_calendar_1,
                            R.drawable.readme_calendar_2,
                        )
                    }

                    is ReadmeSheetVM.Paragraph.AskAQuestion -> {

                        MyListView__ItemView(
                            isFirst = true,
                            isLast = true,
                            modifier = Modifier
                                .padding(top = 24.dp),
                        ) {
                            MyListView__ItemView__ButtonView(text = paragraph.title) {
                                askAQuestion(subject = paragraph.subject)
                            }
                        }
                    }
                }
            }
        }

        Sheet__BottomViewClose {
            layer.close()
        }
    }
}

@Composable
private fun PTitleView(
    text: String,
) {
    Text(
        text = text,
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = hPadding)
            .padding(top = 48.dp),
        color = c.white,
        fontWeight = FontWeight.Bold,
        fontSize = 28.sp,
    )
}

@Composable
private fun PTextView(
    text: String,
    topPadding: Dp = 16.dp,
) {
    Text(
        text = text,
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = hPadding)
            .padding(top = topPadding),
        color = c.white,
        lineHeight = pTextLineHeight,
        fontWeight = FontWeight.Normal,
    )
}

@Composable
private fun PTextHighLightView(
    text: String,
) {
    Text(
        text = text,
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 16.dp)
            .background(c.blue)
            .padding(horizontal = hPadding)
            .padding(top = 12.dp, bottom = 10.dp),
        color = c.white,
        lineHeight = pTextLineHeight,
        fontWeight = FontWeight.Normal,
    )
}

@Composable
private fun PListDashedView(
    items: List<String>,
) {
    VStack {

        items.forEach { item ->

            HStack(
                modifier = Modifier
                    .padding(top = 8.dp),
            ) {

                Icon(
                    painter = painterResource(R.drawable.sf_minus_medium_regular),
                    contentDescription = item,
                    modifier = Modifier
                        .padding(start = hPadding, top = 5.dp)
                        .size(12.dp),
                    tint = c.white,
                )

                Text(
                    text = item,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(start = 12.dp, end = hPadding),
                    color = c.white,
                    lineHeight = pTextLineHeight,
                    fontWeight = FontWeight.Normal,
                )
            }
        }
    }
}

private val imageBorderColor = ColorRgba(96, 96, 96).toColor()
private val imageSliderShape = SquircleShape(len = 90f, angleParam = 10f)

private val imageSliderEnterAnimation: EnterTransition = slideInVertically(
    animationSpec = spring(
        stiffness = Spring.StiffnessMedium,
        visibilityThreshold = IntOffset.VisibilityThreshold,
    ),
    initialOffsetY = { it },
)

private val imageSliderExitAnimation: ExitTransition = slideOutVertically(
    animationSpec = spring(
        stiffness = Spring.StiffnessMedium,
        visibilityThreshold = IntOffset.VisibilityThreshold,
    ),
    targetOffsetY = { it },
)

@Composable
private fun ImagePreviewsView(
    vararg resIds: Int,
    paddingTop: Dp = 20.dp,
) {
    val scrollState = rememberScrollState()
    HStack(
        modifier = Modifier
            .padding(top = paddingTop, bottom = 8.dp)
            .padding(horizontal = imagesHBlock)
            .horizontalScroll(scrollState),
    ) {

        resIds.forEach { resId ->

            Image(
                painter = painterResource(resId),
                modifier = Modifier
                    .height(250.dp)
                    .padding(horizontal = imagesHBetween)
                    .clip(imagesShape)
                    .border(1.dp, imageBorderColor, shape = imagesShape)
                    .clickable {
                        showImagesSlider(*resIds)
                    },
                contentDescription = "Screenshot",
                contentScale = ContentScale.Fit,
            )
        }
    }
}

private fun showImagesSlider(
    vararg resIds: Int,
) {

    WrapperView.Layer(
        enterAnimation = imageSliderEnterAnimation,
        exitAnimation = imageSliderExitAnimation,
        alignment = Alignment.BottomCenter,
        onClose = {},
        content = { layer ->
            VStack(
                modifier = Modifier
                    .background(c.black)
                    .fillMaxSize()
                    .padding(top = statusBarHeight)
                    .pointerInput(Unit) { },
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {

                val scrollState = rememberScrollState()
                HStack(
                    modifier = Modifier
                        .weight(1f)
                        .horizontalScroll(scrollState),
                ) {

                    Padding(horizontal = 8.dp)

                    resIds.forEach { resId ->
                        Image(
                            painter = painterResource(resId),
                            modifier = Modifier
                                .padding(horizontal = 8.dp)
                                .fillMaxHeight()
                                .clip(imageSliderShape)
                                .border(1.dp, imageBorderColor, shape = imageSliderShape),
                            contentDescription = "Screenshot",
                            contentScale = ContentScale.FillHeight,
                        )
                    }

                    Padding(horizontal = 8.dp)
                }

                Text(
                    text = "Close",
                    modifier = Modifier
                        .padding(end = 16.dp)
                        .padding(vertical = 8.dp)
                        .align(Alignment.End)
                        .navigationBarsPadding()
                        .clip(roundedShape)
                        .clickable {
                            layer.close()
                        }
                        .padding(horizontal = 14.dp)
                        .padding(top = 6.dp, bottom = 7.dp),
                    color = c.textSecondary,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Light,
                )
            }
        }
    ).show()
}
